File: 0039-combination-sum/0039-combination-sum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(candidates, target, start, current_combination, result):
            if target == 0:
                result.append(list(current_combination))
                return
            
            for i in range(start, len(candidates)):
                if candidates[i] > target:
                    break
                
                current_combination.append(candidates[i])
                dfs(candidates, target - candidates[i], i, current_combination, result)
                current_combination.pop()

        candidates.sort()
        result = []
        dfs(candidates, target, 0, [], result)
        
        return result

File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let result = [];
    let currentCombination = [];
    
    candidates.sort((a, b) => a - b);
    dfs(candidates, target, 0, currentCombination, result);
    
    return result;
};

var dfs = function(candidates, target, start, currentCombination, result) {
    if (target === 0) {
        result.push(Array.from(currentCombination));
        return;
    }
    
    for (let i = start; i < candidates.length && candidates[i] <= target; i++) {
        currentCombination.push(candidates[i]);
        dfs(candidates, target - candidates[i], i, currentCombination, result);
        currentCombination.pop();
    }
};
